### PR TITLE
Build wheels in parallel, add aarch64

### LINKED ISSUE #3061: ENH: Wheels for aarch64
### Problem Description

Please make wheels for linux arm. context, i'm trying to add shap inside airflow container. i'm using m1 with docker
preferably for python 3.8
```
#0 46.07 Building wheels for collected packages: shap, crcmod, docopt
#0 46.07   Building wheel for shap (pyproject.toml): started
#0 46.60   Building wheel for shap (pyproject.toml): finished with status 'error'
#0 46.62   Building wheel for crcmod (setup.py): started
#0 46.62   error: subprocess-exited-with-error
#0 46.62   
#0 46.62   × Building wheel for shap (pyproject.toml) did not run successfully.
#0 46.62   │ exit code: 1
#0 46.62   ╰─> [123 lines of output]
#0 46.62       The nvcc binary could not be located in your $PATH. Either  add it to your path, or set $CUDAHOME to enable CUDA
#0 46.62       Error building cuda module: TypeError('cannot unpack non-iterable NoneType object')
#0 46.62       WARNING: Could not compile cuda extensions
#0 46.62       running bdist_wheel
#0 46.62       running build
#0 46.62       running build_py
#0 46.62       creating build
#0 46.62       creating build/lib.linux-aarch64-cpython-38
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap
#0 46.62       copying shap/datasets.py -> build/lib.linux-aarch64-cpython-38/shap
#0 46.62       copying shap/links.py -> build/lib.linux-aarch64-cpython-38/shap
#0 46.62       copying shap/_serializable.py -> build/lib.linux-aarch64-cpython-38/shap
#0 46.62       copying shap/_explanation.py -> build/lib.linux-aarch64-cpython-38/shap
#0 46.62       copying shap/__init__.py -> build/lib.linux-aarch64-cpython-38/shap
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_gradient.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/tf_utils.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_exact.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_explainer.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/pytree.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_permutation.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_additive.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_sampling.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_gpu_tree.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_partition.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/mimic.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_tree.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_linear.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       copying shap/explainers/_kernel.py -> build/lib.linux-aarch64-cpython-38/shap/explainers
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/_treegain.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/_lime.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/_random.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/_maple.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/_coefficent.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       copying shap/explainers/other/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/other
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/explainers/_deep
#0 46.62       copying shap/explainers/_deep/deep_tf.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/_deep
#0 46.62       copying shap/explainers/_deep/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/_deep
#0 46.62       copying shap/explainers/_deep/deep_pytorch.py -> build/lib.linux-aarch64-cpython-38/shap/explainers/_deep
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_labels.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_benchmark.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_heatmap.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_force_matplotlib.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_text.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_image.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_utils.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_group_difference.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_waterfall.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_violin.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_force.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_bar.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_beeswarm.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_scatter.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_decision.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_monitoring.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_partial_dependence.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       copying shap/plots/_embedding.py -> build/lib.linux-aarch64-cpython-38/shap/plots
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/plots/colors
#0 46.62       copying shap/plots/colors/_colorconv.py -> build/lib.linux-aarch64-cpython-38/shap/plots/colors
#0 46.62       copying shap/plots/colors/_colors.py -> build/lib.linux-aarch64-cpython-38/shap/plots/colors
#0 46.62       copying shap/plots/colors/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/plots/colors
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/models.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/metrics.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/measures.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/framework.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/plots.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/_compute.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/_result.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/_explanation_error.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/_sequential.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/experiments.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       copying shap/benchmark/methods.py -> build/lib.linux-aarch64-cpython-38/shap/benchmark
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_masker.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_text.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_image.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_fixed.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_tabular.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_fixed_composite.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_output_composite.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/_composite.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       copying shap/maskers/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/maskers
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_clustering.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/image.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_show_progress.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_general.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_keras.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_masked_model.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_exceptions.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/_legacy.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/transformers.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       copying shap/utils/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/utils
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/actions
#0 46.62       copying shap/actions/_action.py -> build/lib.linux-aarch64-cpython-38/shap/actions
#0 46.62       copying shap/actions/_optimizer.py -> build/lib.linux-aarch64-cpython-38/shap/actions
#0 46.62       copying shap/actions/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/actions
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/_teacher_forcing.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/_model.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/_topk_lm.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/_text_generation.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/_transformers_pipeline.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       copying shap/models/__init__.py -> build/lib.linux-aarch64-cpython-38/shap/models
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/plots/resources
#0 46.62       copying shap/plots/resources/bundle.js -> build/lib.linux-aarch64-cpython-38/shap/plots/resources
#0 46.62       copying shap/plots/resources/logoSmallGray.png -> build/lib.linux-aarch64-cpython-38/shap/plots/resources
#0 46.62       creating build/lib.linux-aarch64-cpython-38/shap/cext
#0 46.62       copying shap/cext/tree_shap.h -> build/lib.linux-aarch64-cpython-38/shap/cext
#0 46.62       running build_ext
#0 46.62       numpy.get_include() /tmp/pip-build-env-y3wavfj9/overlay/lib/python3.8/site-packages/numpy/core/include
#0 46.62       building 'shap._cext' extension
#0 46.62       creating build/temp.linux-aarch64-cpython-38
#0 46.62       creating build/temp.linux-aarch64-cpython-38/shap
#0 46.62       creating build/temp.linux-aarch64-cpython-38/shap/cext
#0 46.62       gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include/python3.8 -I/tmp/pip-build-env-y3wavfj9/overlay/lib/python3.8/site-packages/numpy/core/include -c shap/cext/_cext.cc -o build/temp.linux-aarch64-cpython-38/shap/cext/_cext.o
#0 46.62       error: command 'gcc' failed: Permission denied
#0 46.62       [end of output]
#0 46.62   
#0 46.62   note: This error originates from a subprocess, and is likely not a problem with pip.
#0 46.62   ERROR: Failed building wheel for shap
```

### Alternative Solutions

_No response_

### Additional Context

_No response_

### Feature request checklist

- [X] I have checked the issue tracker for duplicate issues.
- [ ] I'd be interested in making a PR to implement this feature

### PR DESCRIPTION
**Pull request by connortann**
This PR should prepare us for a patch release `0.42.1`. I've bumped the version number and changelog accordingly.

Changes:

- Modifies our wheel job to run in parallel, which is must faster and makes the jobs more understandable.
- Changes the cibuildwheel dependencies from `test` to `test-core`, to only run the limited test set on the built wheels.
- Adds `aarch64` on linux, closes #3061
- Skips tests on emulated architectures (which are very slow)

Example [wheel build workflow run](https://github.com/slundberg/shap/actions/runs/5540996538) (passes).

Example job matrix:

image

### CODE REVIEW

**Comment by connortann** [comment]
@thatlittleboy if you'd kindly give a review, I can go ahead and publish the patch release once this PR is merged.

**Comment by connortann** [comment]
@PrimozGodec  let me know if you have any comments on this PR. I appreciate your input so far regarding wheels; let me know if there any other commonly used architectures that you think we should include on top of what is currently in `pyproject.toml`.

**Comment by PrimozGodec** [comment]
Thank you for the pull request and all the work on Shap (and toward the patch release). For me, it looks ok. If I compare wheels offered with Scipy/Numpy, only musllinux wheels are missing (which are [mostly used in containers](https://www.youtube.com/watch?v=QAbveJB5kk8&ab_channel=anthonywritescode)).

I am trying to make them available at 
https://github.com/slundberg/shap/pull/3087, but it is not as crucial as other weels the new release will offer.

All other wheels are unnecessary since users must build Numpy and Scipy anyway. I also noticed that cp37 and cp38 wheels could also be dropped since Numpy and Scipy dropped them already (as I understood it was planned to be done in the future).

(Sorry for the late reply; I am currently on vacation).

**Comment by connortann** [comment]
That's helpful, thanks. Enjoy your vacation - hope you resist the urge to check GitHub!

**Comment by PrimozGodec** [comment]
I would say that the release is made without musllinux wheels, and I can work on them when I am back. Tests do not go through because some packages (scikit-learn) do not offer musllinux wheels and would need to be properly compiled.

> That's helpful, thanks. Enjoy your vacation - hope you resist the urge to check GitHub!

It was no problem, happy to help. :)

**Comment by connortann** [comment]
Created an issue to track the musllinux proposal for future reference: #3089 .

**Review by thatlittleboy — APPROVED** [review]
@connortann sorry for the delay. let's try to get this out this weekend if possible.

borat-great-success
